crates/redis/src/lib.rs

```diff
@@ -52,30 +55,81 @@ impl TriggerExecutor for RedisTrigger {
     type RunConfig = NoArgs;
 
     async fn new(engine: TriggerAppEngine<Self>) -> Result<Self> {
-        let address = engine
+        let default_address: String = engine
             .trigger_metadata::<TriggerMetadata>()?
             .unwrap_or_default()
             .address;
 
+        let mut addresses = vec![default_address.clone()];
+
         let mut channel_components: HashMap<String, Vec<String>> = HashMap::new();
 
         for (_, config) in engine.trigger_configs() {
+            let address = config.address.clone().unwrap_or(default_address.clone());
+            if !addresses.contains(&address) {
+                addresses.push(address.clone());
+            }
             channel_components
-                .entry(config.channel.clone())
+                .entry(format!("{}:{}", address, config.channel.clone()))
                 .or_default()
                 .push(config.component.clone());
         }
         Ok(Self {
-            engine,
-            address,
+            engine: Arc::new(engine),
+            addresses,
             channel_components,
         })
     }
 
     /// Run the Redis trigger indefinitely.
     async fn run(self, _config: Self::RunConfig) -> Result<()> {
-        let address = &self.address;
+        let addresses = self.addresses.clone();
+        let mut tasks = Vec::new();
+        // Spawn an asynchronous task for each iteration
+        for address in addresses {
+            let trigger = self.clone();
+            // Spawn an asynchronous task with a closure capturing self by reference
+            println!("Starting listener for {address}");
+            tasks.push(tokio::spawn(
+                async move { trigger.run_listener(address).await },
+            ));
+        }
+
+        let _ = join_all(tasks).await;
```

Comment:
I think we should possibly exit when one of the listeners exits. thoughts?

Reply:
Either exit so that the alarm goes off, or try to restart the relevant listener.  But yeah, I don't think we want it continuing in a partial state.

(For what it's worth I think the SQS trigger exits, but that may be because I am lazy.)

Reply:
Looking at the SQS trigger `futures::future::select_all` seems to be what I thought would work